File: js/euler.js
// Project Euler: JavaScript Solutions

// Problem 1: Multiples of 3 and 5
function sumOfMultiples(number) {
  var sumOfThree = 0;
  var sumOfFive = 0;
  var sumTotal = 0;

  if (number < 5) {
    sumOfFive = 0;
  } else {
    for (var i = 5; i < number; i += 5) {
      sumOfFive += i;
    }
  }

  if (number < 3) {
    sumOfThree = 0;
  } else {
    for (var j = 3; j < number; j += 3) {
      // Check for mutiples of 5 already counted.
      if (j % 5 != 0) {
        sumOfThree += j;
      }
    }
  }

  sumTotal = sumOfFive + sumOfThree;

  return sumTotal;
}

function problem1() {
  document.getElementById("answer1").innerHTML = 
    sumOfMultiples(document.getElementById("problem1").value);
}

// Problem 2: Even Fibonacci Numbers
function evenFibonacci(number) {
  var evenSum = 0;
  var fib1 = 1;
  var fib2 = 2;
  var fibCount = 0;

  // Base case
  if (number < 3) {
    evenSum = 0;
  } else {
    while (fibCount < number) {
      if (fib2 % 2 === 0) {
        evenSum += fib2;
      }
      fibCount = fib1 + fib2;
      fib1 = fib2;
      fib2 = fibCount;
    }
  }

  return evenSum;
}

function problem2() {
  document.getElementById("answer2").innerHTML = 
    evenFibonacci(document.getElementById("problem2").value);
}

// Problem 3: Largest Prime Factor
function largestPrime(number) {
  var isPrime = false;
  var currentNum = number;
  var divisor = 2;
  var maxPrime = 1;

  while (isPrime != true && divisor <= currentNum) {
    if (currentNum % divisor === 0) {
      currentNum /= divisor;
      if (divisor > maxPrime) {
        maxPrime = divisor;
      }
      divisor = 2;
    } else {
      divisor++;
    }

    if (divisor > currentNum) {
      isPrime = true;
    }
  }
  return maxPrime;
}

function problem3() {
  document.getElementById("answer3").innerHTML = 
    largestPrime(document.getElementById("problem3").value);
}

File: ruby/euler.rb
# Project Euler: Ruby Solutions

# Problem 1: Multiples of 3 and 5
def sum_of_multiples(number)
  sum_three = 0
  sum_five = 0
  sum_total = 0
  count_three = 3
  count_five = 5

  if number < 5
    sum_five
  else
    while count_five < number
      sum_five += count_five
      count_five += 5
    end
  end

  if number < 3
    sum_three
  else
    while count_three < number
      if count_three % 5 != 0
        sum_three += count_three
      end
      count_three += 3
    end
  end

  sum_total = sum_five + sum_three

  puts sum_total
  return sum_total
end

sum_of_multiples(1000)

# Problem 2: Even Fibonacci Numbers
def even_fibonacci(number)
  even_sum = 0
  fib1 = 1
  fib2 = 2
  fib_count = 0

  if number < 3
    even_sum = 0
  else
    while fib_count < number
      if fib2 % 2 === 0
        even_sum += fib2
      end
      fib_count = fib1 + fib2
      fib1 = fib2
      fib2 = fib_count
    end
  end

  puts even_sum
  return even_sum
end

even_fibonacci(4000000)

# Problem 3: Largest Prime Factor
def largest_prime(number)
  is_prime = false
  current_num = number
  divisor = 2
  max_prime = 1

  while is_prime != true && divisor <= current_num
    if current_num % divisor === 0
      current_num /= divisor
      if divisor > max_prime
        max_prime = divisor
      end
      divisor = 2
    else
      divisor += 1
    end

    if divisor > current_num
      is_prime = true
    end
  end

  puts max_prime
  return max_prime
end

largest_prime(600851475143)